RFC: ADC: add support for injected channels for STM32

gautierg-st opened the pull request:
This PR adds the support for the STM32 ADC injected channels. Starting an injected sequence means that the sequence is executed immediately. If a regular sequence is being executed, it will be paused and resumed once the injected one is done.

It makes a slight change to the ADC API by adding a boolean in the `adc_sequence` structure to indicate if the sequence should be injected or not. This boolean is itself hidden behind a specific Kconfig. That way, there is no difference for ADC that do not support this feature.

An ADC test is added to test the feature on an STM32U5 board. The PR is functional but the implementation stays open for discussion.

gautierg-st committed:
include: zephyr: drivers: adc: add sequence priority

Add sequence priority support in the ADC API.
The priority assigned to a sequence reflects how time-sensitive it is
compared to other sequences. The number of priority levels supported
and the exact behavior obtained by assigning a specific priority to a
sequence are hardware-specific.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: stm32: add injected mode support

Add injected mode support for STM32 ADC driver. To avoid potential
conflicts when configuring the ADC registers, some settings (like presel)
are moved into the channel_setup function when injected is enabled.

Also configures an ISR when DMA and injected mode are enabled. The reason
is that DMA is not available for injected sequence, so an ISR is still
necessary in this case.

Factorizes the "start_read_xxx" functions to avoid needless duplication
between sync/async and regular/injected mode.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: sensor: stm32_temp: add injected mode support

Add injected mode support for STM32 temperature sensor.
When in injected mode, this configures the sensor channel and sensor enable
bit at start to prevent any conflict later: some registers can't be written
while a conversion is on-going.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: sensor: stm32_vbat: add injected mode support

Add injected mode support for STM32 Vbat sensor.
When in injected mode, this configures the sensor channel and sensor enable
bit at start to prevent any conflict later: some registers can't be written
while a conversion is on-going.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: sensor: stm32_vref: add injected mode support

Add injected mode support for STM32 Vref sensor.
When in injected mode, this configures the sensor channel and sensor enable
bit at start to prevent any conflict later: some registers can't be written
while a conversion is on-going.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
tests: drivers: adc_api: add a test for injected channels

Add a test for injected channels. The test first starts a regular
asynchronous sampling of several channels, then starts an injected
sampling. The test checks that both the regular and injected samplings
run without issues.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

etienne-lms commented on drivers/adc/Kconfig:
```diff
@@ -38,6 +38,12 @@ config ADC_CONFIGURABLE_EXCITATION_CURRENT_SOURCE_PIN
+# support injected channels, i.e. reading a sequence of channels outside of the
+# main sequence.
+config ADC_INJECTED_CHANNELS
+	bool
```
Could you more the description in the `help` section?

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -1145,6 +1275,23 @@ static void adc_stm32_isr(const struct device *dev)
+		LL_ADC_ClearFlag_JEOS(adc);
+		for (uint8_t i = 0; i < data->channel_count_injected; i++) {
+			*data->buffer_injected++ = LL_ADC_INJ_ReadConversionData32(adc,
+						   table_rank_injected[i]);
```
Could you add an indentation tab?

etienne-lms commented on drivers/adc/Kconfig:
```diff
@@ -38,6 +38,12 @@ config ADC_CONFIGURABLE_EXCITATION_CURRENT_SOURCE_PIN
+# By selecting or not this option particular ADC drivers indicate if they
+# support injected channels, i.e. reading a sequence of channels outside of the
+# main sequence.
+config ADC_INJECTED_CHANNELS
```
Deserve an add in the release note?

mathieuchopstm commented on drivers/adc/adc_stm32.c:
```diff
@@ -1405,6 +1583,19 @@ static int adc_stm32_channel_setup(const struct device *dev,
+
+		LL_ADC_SetChannelPreselection(adc, channel);
+	}
+#endif /* ANY_ADC_HAS_CHANNEL_PRESELECTION */
```
Not sure how this change relates to injected channels. Maybe it should be moved to a dedicated commit?

gautierg-st replied:
Writing to the preselection register cannot be done if any sequence (regular or injected) is on-going. So it is necessary to set it up at the "channel setup" stage rather than at the start stage. That's also why this register is not reset when sampling is done (when injected is enabled).

mathieuchopstm commented on drivers/sensor/st/stm32_temp/Kconfig:
```diff
@@ -14,3 +14,13 @@ config STM32_TEMP
+	bool "STM32 Temperature Sensor through injected channel"
+	select ADC_STM32_INJECTED_CHANNELS
+	help
+	  Temperature measurement is done using the injected channel feature.
```
Add a note that this enables the sensor for entire application lifetime, which may be unwanted by some applications. Ditto other drivers.

erwango commented on drivers/adc/adc_stm32.c:
```diff
@@ -1995,6 +2196,12 @@ DT_INST_FOREACH_STATUS_OKAY(GENERATE_ISR)
+		IF_ENABLED(CONFIG_ADC_STM32_INJECTED_CHANNELS,					\
+			(ADC_CONTEXT_INIT_LOCK(adc_stm32_data_##index, ctx_injected),))		\
+		IF_ENABLED(CONFIG_ADC_STM32_INJECTED_CHANNELS,					\
+			(ADC_CONTEXT_INIT_SYNC(adc_stm32_data_##index, ctx_injected),))		\
```
Can it be factorized to ease reading ? Sthg like:
```
		IF_ENABLED(CONFIG_ADC_STM32_INJECTED_CHANNELS,					\
			(ADC_CONTEXT_INIT_TIMER(adc_stm32_data_##index, ctx_injected),)	\
			(ADC_CONTEXT_INIT_LOCK(adc_stm32_data_##index, ctx_injected),)		\
			(ADC_CONTEXT_INIT_SYNC(adc_stm32_data_##index, ctx_injected),))		\
```

mathieuchopstm commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -38,6 +38,10 @@ properties:
 
+  "#io-injected-channel-cells":
+    type: int
+    const: 1
```
There should be a description somewhere for what these are, unless this is also standard like `io-channels` is

mathieuchopstm reviewed (commented):
Seems overall OK; some comments:

mathieuchopstm commented on drivers/adc/adc_stm32.c:
```diff
@@ -991,6 +1025,91 @@ static int set_sequencer(const struct device *dev)
+	 * - channel_id: ordinal position of channel in data->channels bitmask
+	 */
+	for (uint32_t channels = data->channels_injected; channels;
+		      channels &= ~BIT(channel_id), channel_index++) {
```
\+1 but keep the `channels != 0` inside for loop condition

mathieuchopstm commented on drivers/sensor/st/stm32_temp/stm32_temp.c:
```diff
@@ -276,8 +280,22 @@ static int stm32_temp_init(const struct device *dev)
+	}
+
+	adc_enable_tempsensor_channel(data->adc_base);
+#endif /* CONFIG_STM32_TEMP_INJECTED */
```
Oh, right, the settings set by `adc_channel_setup` only apply to the particular channel, so they won't be overriden by another caller (unless as you said, the sensor channel is specified, but then it is an error from that caller).

Resolving conversation.

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -951,16 +995,6 @@ static int set_sequencer(const struct device *dev)
 
 #if ANY_ADC_SEQUENCER_TYPE_IS(SEQUENCER_PROGRAMMABLE)
 		if (config->sequencer_type == SEQUENCER_PROGRAMMABLE) {
-#if ANY_ADC_HAS_CHANNEL_PRESELECTION
```
Writing to the preselection register cannot be done if any sequence (regular or injected) is on-going. So it is necessary to set it up at the "channel setup" stage rather than at the start stage. That's also why this register is not reset when sampling is done (when injected is enabled).

That said, the code should be kept here when injected is not enabled, or else some current tests/samples will fail (because presel will be reset at the end of the sampling, and `channel_setup` is not necessarily called each time before starting conversions)

gautierg-st committed:
drivers: adc: stm32: do not stop conversion unconditionally

In order to write into some ADC registers, it is necessary to have no
on-going conversions (either regular, injected, or both). Until now, each
time such a write was needed, the ADC was disabled unconditionally,
stopping any potential on-going conversion.

With the introduction of the injected mode, this could lead to problems
if for example an injected is configured differently than a running
regular one. In that case it would stop the regular conversion. If the
application is waiting the result before starting another one, it will be
stuck waiting forever.

To avoid cases like this, return an error rather than stopping conversion
and disabling the ADC. This reworks many functions to return an error in
case it was not possible to safely disable the ADC.

It stays possible to force an ADC disable even though a conversion is
running, this is useful at init where we don't want this to fail.

This also creates a dedicated function for the preselection register to
avoid duplication.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
doc: releases: 4.4: add the new api for adc sequence priority

The ADC has a new API to enable sequence priority
(Kconfig + struct member).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st commented on drivers/adc/Kconfig:
```diff
@@ -38,6 +38,12 @@ config ADC_CONFIGURABLE_EXCITATION_CURRENT_SOURCE_PIN
 
+# By selecting or not this option particular ADC drivers indicate if they
+# support injected channels, i.e. reading a sequence of channels outside of the
+# main sequence.
```
Reworked with only one Kconfig. Let me know what you think

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -208,6 +233,14 @@ struct adc_stm32_data {
+	uint32_t channels_injected;
+	uint8_t channel_count_injected;
+	uint8_t samples_count_injected;
+#endif /* CONFIG_ADC_STM32_INJECTED_CHANNELS */
```
Done

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -1216,6 +1379,20 @@ static int adc_stm32_read_async(const struct device *dev,
+
+		return error;
+	}
+#endif /* CONFIG_ADC_STM32_INJECTED_CHANNELS */
```
Done

gautierg-st commented on drivers/sensor/st/stm32_vbat/stm32_vbat.c:
```diff
@@ -34,12 +34,27 @@ struct stm32_vbat_config {
 	int ratio;
 };
 
+static inline void adc_enable_vbatsensor_channel(ADC_TypeDef *adc)
```
Done. (Though I merely followed the pattern that you initiated in the temp sensor ;) )

gautierg-st commented on drivers/sensor/st/stm32_vref/stm32_vref.c:
```diff
@@ -38,12 +38,31 @@ struct stm32_vref_config {
 	uint8_t cal_shift;
 };
 
+static inline void adc_enable_vrefsensor_channel(ADC_TypeDef *adc)
```
Done

gautierg-st commented on tests/drivers/adc/adc_api/src/test_adc.c:
```diff
@@ -97,19 +121,30 @@ static void init_adc(void)
 	TC_PRINT("Samples read: ");
-	for (i = 0; i < BUFFER_SIZE; i++) {
-		adc_data_size_t sample_value = m_sample_buffer[i];
+	for (i = 0; i < MAX(BUFFER_SIZE, expected_count); i++) {
```
Done

gautierg-st commented:
Last push was to fix conflicts for the release notes.

mathieuchopstm reviewed (commented):
Otherwise LGTM

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -1527,6 +1605,7 @@ static int adc_stm32_channel_setup(const struct device *dev,
 #if defined(CONFIG_SOC_SERIES_STM32H5X) || ANY_ADC_HAS_DIFFERENTIAL_SUPPORT || \
 	(ANY_ADC_HAS_CHANNEL_PRESELECTION && CONFIG_ADC_STM32_INJECTED_CHANNELS)
 	ADC_TypeDef *adc = config->base;
+	int err;
```
I think if only `CONFIG_SOC_SERIES_STM32H5X` is true among above conditions, `err` will trigger an unused variable build warning.

etienne-lms commented on drivers/sensor/st/stm32_vref/Kconfig:
```diff
@@ -12,3 +12,14 @@ config STM32_VREF
+
+config STM32_VREF_INJECTED
+	bool "STM32 Vref Sensor through injected channel"
+	depends on ADC_INJECTED_CHANNELS && ADC_STM32_INJECTED_CHANNELS
```
```suggestion
	depends on ADC_STM32_INJECTED_CHANNELS

gautierg-st replied:
Done

etienne-lms reviewed (commented):
Some minor comments. No blocking issues.

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -763,43 +786,48 @@ static const uint32_t table_oversampling_ratio[] = {
-	 * For the G0 series, ADC must be disabled to prevent CKMODE bitfield
-	 * from getting reset, see errata ES0418 section 2.6.4.
+	 * setting OVS bits is conditioned to ADC state: ADC must be disabled
+	 * or enabled without conversion on going : disable it, it will stop
```
Done

erwango reviewed (changes requested):
Nitpick comment.
Otherwise, it would be nice to have feeback on the API of the test changes.
@MaureenHelm Is that fine for you  ?

gautierg-st commented on drivers/adc/adc_stm32.c:
```diff
@@ -455,43 +455,63 @@ static void adc_stm32_start_conversion(const struct device *dev)
 /*
  * Disable ADC peripheral, and wait until it is disabled
```
Done

etienne-lms reviewed (changes requested):
A few typos in `CONFIG_ADC_STM32_INJECTED_CHANNELS` directive tests.

The SonarCloud reported issues look like false positive to me.

etienne-lms reviewed (commented):
LGTM aside the 2 issues in the first transitory commit.

MaureenHelm reviewed (commented):
Maybe I missed it, but has this been presented at the Arch WG?

erwango commented on include/zephyr/drivers/adc.h:
```diff
@@ -692,6 +692,15 @@ struct adc_sequence {
+	 *
+	 * If true, the sequence will be sampled using the injected mode
+	 */
+	bool injected_mode;
```
@MaureenHelm, fixed by latest push accoridng to this proposal: https://github.com/zephyrproject-rtos/zephyr/pull/99787#issuecomment-3784432631

etienne-lms commented on drivers/adc/adc_stm32.c:
```diff
@@ -390,6 +422,23 @@ static int adc_stm32_enable(ADC_TypeDef *adc)
+	LOG_DBG("Starting injected conversion");
+
+#if !defined(CONFIG_SOC_SERIES_STM32F1X) && \
+	!DT_HAS_COMPAT_STATUS_OKAY(st_stm32f4_adc)
```
```suggestion
#if !DT_HAS_COMPAT_STATUS_OKAY(st_stm32f1_adc) && !DT_HAS_COMPAT_STATUS_OKAY(st_stm32f4_adc)
```
for stm32f3x? (**stm32f373.dtsi** relies on `st,stm32f1-adc` compatible)

gautierg-st replied:
You're right that F373 relies on `st;stm32f1-adc` compatible, but in this particular case, the F373 uses the same function name as the other F3x (and most other series) and not the F1x/F4x name. So nothing to change here.

gautierg-st committed:
dts: bindings: adc: stm32: add property to indicate injected support

Add a property to the STM32 ADC bindings to indicate whether the ADC
instance supports the injected channel feature.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

gautierg-st committed:
drivers: adc: use the new injected support property in the driver

Use the new has-injected-support property in the Kconfig instead of
filtering on series directly, and use it also in the driver to return an
error if a specific instance doesn't support injected (this is the case
for STM32U5 where ADC1 and 2 support injected, but not ADC4).

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>

mathieuchopstm commented on tests/drivers/adc/adc_api/boards/b_u585i_iot02a.overlay:
```diff
@@ -7,18 +7,122 @@
+		/* high-prio-indexes indicates which channels (from io-channels) will be used with a
+		 * higher priority. It indicates the index in the phandle, not the channel value.
+		 */
+		high-prio-indexes = <4>, <9>, <14>;
```
Alternative idea: IIRC you can define nodes in `zephyr,user`, so maybe you could have something like
```dts
io-channels = <...>;

/* child nodes are used to test additional priorities */
injected {
   priority = <1>;
   io-channels = </* injected channels list */>;
};
```

The test can then iterate the child nodes of `zephyr,user` and create additional test cases based on that. This also has the advantage of scaling for more than 1 priorities, and using different channels for the "normal" and "prioritized" tests which some hardware might require (🤷‍♂️)

Roughly:
```c
#define _PRIO_CHANNELS_DEFINE(node, prio)       \
    static const struct adc_dt_spec CONCAT(adc_channels_prio_, prio)[] = { \
	DT_FOREACH_PROP_ELEM(node, io_channels, DT_SPEC_AND_COMMA)         \
    };

#define PRIO_CHANNELS_DEFINE(node)                                        \
    COND_CODE_1(DT_NODE_HAS_PROP(node, priority),                         \
        (_PRIO_CHANNELS_DEFINE(node, DT_PROP(node, priority))),           \
        (BUILD_ASSERT(0, /* some error */);))

DT_FOREACH_CHILD(DT_PATH(zephyr_user), PRIO_CHANNELS_DEFINE);

struct prio_test_case {
    int prio;
    const struct adc_dt_spec *channels;
};

#define _PRIO_TESTCASE_DEFINE(prio) \
        { prio, &CONCAT(adc_channels_prio_, prio) }

#define PRIO_TESTCASE_DEFINE(node)  \
        IF_ENABLED(DT_NODE_HAS_PROP(node, priority),                         \
          (_PRIO_TESTCASE_DEFINE(node, DT_PROP(node, priority))))

static const struct prio_test_case prio_test_cases[] = {
    DT_FOREACH_CHILD_SEP(DT_PATH(zephyr_user), PRIO_TESTCASE_DEFINE, (,))
    /* may need a NULL last such that array is not empty
     * or gate behind DT_CHILD_NUM(DT_PATH(zephyr_user)) > 0
     */
};


testcase_adc_priorities() {
    if (ARRAY_SIZE(prio_test_cases) == 0 /* or DT_CHILD_NUM() */) {
       return TC_SKIP;
    }

    for (tc in prio_test_cases) {
       /* run a sequence with tc->channels,
        * setting seq.priority = tc->prio
        */
    }
}
```

gautierg-st commented on dts/bindings/adc/st,stm32-adc.yaml:
```diff
@@ -156,5 +156,10 @@ properties:
+  has-injected-support:
+    type: boolean
+    description: |
+      If present, it indicates that the ADC instance supports the injected channel feature.
```
Thanks. Done

bjarki-andreasen commented on include/zephyr/drivers/adc.h:
```diff
@@ -760,6 +760,15 @@ struct adc_sequence {
 
+#if defined(CONFIG_ADC_SEQUENCE_PRIORITY) || defined(__DOXYGEN__)
+	/**
+	 * Channel priority for arbitration, HW specific
```
number of supported priorities depends on the HW, but not the meaning of the priority value

bjarki-andreasen reviewed (changes requested):
Maybe a bit pedantic but some of the commit messages for the added adc sequence prio reference injected channels rather than sequence prio

mathieuchopstm commented on tests/drivers/adc/adc_api/boards/b_u585i_iot02a.overlay:
```diff
@@ -7,18 +7,124 @@
+		/* io-channels-prio indicates the priority of each channel defined in io-channels
+		 * for test cases with multiple priorities.
+		 */
+		io-channels-prio = <0>, <0>, <0>, <0>, <1>,
```
> We don't have to use the name `io-channels`, which has a rather common usage pattern, we can create our own which matches what the test is doing.

I'm not sure this will work because the binding for `/zephyr,user` is automatically derived, and I'm not sure `gen_defines` can infer `specifier-space: io-channel` from a mere `seq-0 = <&adc1 1>`.

(we could create a test-specific binding / use children nodes, c.f. https://github.com/zephyrproject-rtos/zephyr/pull/99787#discussion_r2773380018)

gautierg-st replied:
Thanks @bjarki-andreasen for this PR. That should indeed solve the problem. I'll wait for it to be merged to update the tests on my PR.

etienne-lms reviewed (commented):
Implementation looks consistent to me. 2 minor comments.

gautierg-st committed:
dts: arm: st: add the new injected support property to dtsi

For all STM32 ADC instances that support the injected channel feature, add
the newly created property to indicate that.

Signed-off-by: Guillaume Gautier <guillaume.gautier-ext@st.com>
